### ISSUE TITLE
Custom Model metaclass causing TypecheckAssertionError

### ISSUE THREAD

**Issue by alexhenman** [typeddjango/django-stubs#1799]
# Bug report

## What's wrong

We're having some problems using `django-stubs` where we've customised the `ModelBase` metaclass for models. Appreciate this is probably not a super common use case but suspect this is likely not just going to cause issues for us. Whilst investigating I noticed that wagtail for example also utilise this pattern:

https://github.com/wagtail/wagtail/blob/56c62cc29874b98c863792628fa5cdf40db1589b/wagtail/models/__init__.py#L221-L251

I modified an existing test to demonstrate the failure:

```patch
---
 tests/typecheck/managers/querysets/test_filter.yml | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/tests/typecheck/managers/querysets/test_filter.yml b/tests/typecheck/managers/querysets/test_filter.yml
index 83b14bc8..f6e46d6c 100644
--- a/tests/typecheck/managers/querysets/test_filter.yml
+++ b/tests/typecheck/managers/querysets/test_filter.yml
@@ -136,9 +136,14 @@
         -   path: myapp/models.py
             content: |
                 from django.db import models
-                class Publisher(models.Model):
+                from django.db.models.base import ModelBase
+                class MyModelBase(ModelBase):
                     pass
-                class Blog(models.Model):
+                class MyModel(models.Model, metaclass=MyModelBase):
+                    pass
+                class Publisher(MyModel):
+                    pass
+                class Blog(MyModel):
                     publisher = models.ForeignKey(Publisher, on_delete=models.CASCADE, related_name='blogs')
```

which results in this output running the tests:

```python
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output: 
E   Actual:
E     main:5: error: Cannot resolve keyword 'publisher_id' into field. Choices are: blog, id, publisher  [misc] (diff)
E     main:8: error: Incompatible type for lookup 'publisher': (got "Blog", expected "Union[Publisher, int, None]")  [misc] (diff)
E     main:9: error: Cannot resolve keyword 'publisher_id' into field. Choices are: blog, id, publisher  [misc] (diff)
E     myapp/models:10: error: Incompatible types in assignment (expression has type "ForeignKey[Union[Publisher, Combinable], Publisher]", base class "MyModel" defined the type as "ReverseOneToOneDescriptor[MyModel, Publisher]")  [assignment] (diff)
E   Expected:
E     main:8: error: Incompatible type for lookup 'publisher': (got "Blog", expected "Union[Publisher, int, None]")  [misc] (diff)
E     main:9: error: Incompatible type for lookup 'publisher_id': (got "Blog", expected "Union[str, int]")  [misc] (diff)
E   Alignment of first line difference:
E     E: main:8: error: Incompatible type for lookup 'publisher': (got "Blog", ex...
E     A: main:5: error: Cannot resolve keyword 'publisher_id' into field. Choices...
E             ^
```

We were previously running mypy 1.4.1 with django-stubs 4.2.3 and weren't seeing any problems so I think this is a result of some changes that have recently been introduced.

## How is that should be

If there's a simple way for us to hint that our metaclass behaves the same way as `ModelBase` a workaround would be very helpful!

But it would be great if it was possible to allow some way of customising `ModelBase` whilst keeping the stubs from `django-stubs`

## System information

- OS:
- `python` version: 3.11.3
- `django` version:
- `mypy` version: 1.6.0
- `django-stubs` version: 4.2.5
- `django-stubs-ext` version:

**Comment by flaeppe** [comment]
I suspect this is a real use case of what we've previously tried to explore in #1670 

CC @sobolevn

**Comment by flaeppe** [comment]
I'm not at all sure, but could it be possible to use `typing_extensions.Annotated` as a workaround? Ref: [PEP 593](https://peps.python.org/pep-0593/)

**Comment by javulticat** [comment]
I can confirm the same behavior as reported by @alexhenman in this issue (#1799).  Using any custom metaclass on a model (rather than using the standard metaclass of a Django model: `django.db.models.base.ModelBase`) seems to cause major issues around relationships (such as the fact that adding a `ForeignKey` field to a model should create a `<field.name>_id` attribute on the model's instances, as shown in the example provided [above](https://github.com/typeddjango/django-stubs/issues/1799#issue-1961811749).  If every single model in your Django app is using a "base" custom metaclass (like ours is 🙃), or you are otherwise making use of custom model metaclasses for a significant number of models, then this bug makes it effectively impossible to use `mypy`+`django-stubs` to do any meaningful static type checking of your Django codebase.

_However_, I **can confirm** that PR #1670 by @flaeppe appears to completely solve this bug!  🙌  Due to factors outside of my control (corporate red tape 😕), creating and using a fork of `django-stubs` was not an option for me, but I still was able to confirm the effectiveness of the change by taking the code directly from the PR and using it to monkey-patch the`mypy_django_plugin.main.NewSemanalDjangoPlugin.get_base_class_hook()` method, setting it to have the same implementation it had in PR #1670.  Lo and behold, upon doing so, every single one of the several hundred `mypy` errors we were seeing thanks to our use of a shared base custom metaclass for all of our models are now no longer raised by `mypy`!  🎉

So, would you kindly consider the revival and merge of PR #1670 in order to resolve this issue?  Thankfully, it seems like the tiny one-line change it makes to the `.get_base_class_hook()` method of `NewSemanalDjangoPlugin` will be able to alleviate what is currently a pretty major pain point for us folks who use `mypy`+`django-stubs` for static type checking their Django codebase that uses custom model metaclasses.

Thanks in advance for taking another look at this issue!  🙏  @flaeppe and @sobolevn, I don't want to step on any toes, but I'd be more than happy to help shepherd PR #1670 through the review process as either a contributor or a reviewer if you folks are busy with other things.

In the meantime, for @alexhenman and anyone else struggling with this issue that also can't/doesn't want to maintain their own `django-stubs` fork, putting this at the top of my `settings.py` is how I successfully monkey-patched in the fix from #1670 to get around my corporate red tape problem (although, very unfortunately, use of this monkey-patch required both `mypy` and `django-stubs` to be installed as production dependencies... 🤢):

```python
from mypy.nodes import TypeInfo
from mypy_django_plugin.lib import fullnames
from mypy_django_plugin.main import NewSemanalDjangoPlugin
from mypy_django_plugin.transformers.models import process_model_class

NewSemanalDjangoPlugin.get_base_class_hook = lambda self, fullname: (
    partial(
        process_model_class,
        django_context=self.django_context,
    )
    if (sym := self.lookup_fully_qualified(fullname)) is not None
    and isinstance(
        sym.node,
        TypeInfo,
    )
    and sym.node.metaclass_type is not None
    and (
        sym.node.metaclass_type.type.fullname == fullnames.MODEL_METACLASS_FULLNAME
        or sym.node.metaclass_type.type.has_base(
            fullnames.MODEL_METACLASS_FULLNAME,
        )
    )
    else None
)
```

## System information

- OS:
    - Dev: macOS 14.1.2 [arm64]
    - Prod: Amazon Linux 2 [arm64] (via [public.ecr.aws/lambda/python:3.11-arm64](https://gallery.ecr.aws/lambda/python) container image deployed to AWS Lambda)
- `python` version: 3.11.6
- `django` version: 4.2.7
- `django-stubs` version: 4.2.6
- `django-stubs-ext` version: 4.2.5

**Comment by flaeppe** [comment]
I've "reopened" #1670 as #1863
